File: scala/revenj-core/src/main/scala/net/revenj/TreePath.scala
package net.revenj

import java.util

final class TreePath private(private val value: String, private val parts: Array[String]) {

  override def hashCode: Int = value.hashCode

  override def equals(obj: Any): Boolean = obj.isInstanceOf[TreePath] && obj.asInstanceOf[TreePath].value == this.value || obj.isInstanceOf[String] && obj == this.value

  override def toString: String = value

  def plus(other: TreePath): TreePath = {
    if (other == null) this
    else if (this.value.length == 0) other
    else if (other.value.length == 0) this
    else {
      val values = util.Arrays.copyOf(this.parts, this.parts.length + other.parts.length)
      var i = 0
      while (i < other.parts.length) {
        values(this.parts.length + i) = other.parts(i)
        i += 1
      }
      new TreePath(this.value + "." + other.value, values)
    }
  }

  def isAncestor(other: TreePath): Boolean = {
    if (other == null) false
    else if (this.value == other.value) true
    else this.parts.length < other.parts.length && TreePath.compare(this.parts, other.parts, this.parts.length)
  }

  def isDescendant(other: TreePath): Boolean = {
    if (other == null) false
    else if (this.value == other.value) true
    else this.parts.length > other.parts.length && TreePath.compare(this.parts, other.parts, other.parts.length)
  }
}

object TreePath {
  val Empty: TreePath = new TreePath("", new Array[String](0))

  def create(value: String): TreePath = {
    if (value == null || value.isEmpty) Empty
    else {
      val parts = value.split("\\.")
      checkParts(parts)
      new TreePath(value, parts)
    }
  }

  private def checkParts(parts: Array[String]): Unit = {
    var i = 0
    while(i < parts.length) {
      val p = parts(i)
      i += 1
      var j = 0
      while (j < p.length) {
        if (!Character.isLetterOrDigit(p.charAt(j))) throw new IllegalArgumentException("Invalid value for part: " + p + ". Only letters and digits allowed for labels")
        j += 1
      }
    }
  }

  private def compare(left: Array[String], right: Array[String], count: Int): Boolean = {
    var i = 0
    var same = true
    while (same && i < count) {
      same = left(i) == right(i)
      i += 1
    }
    same
  }
}

File: java/revenj-core/src/main/java/org/revenj/TreePath.java
package org.revenj;

import java.util.Arrays;

public final class TreePath {
	private final String value;
	private final String[] parts;

	public static final TreePath EMPTY = new TreePath("", new String[0]);

	public TreePath(String value) {
		if (value == null || value.length() == 0) {
			this.value = "";
			this.parts = new String[0];
		} else {
			this.value = value;
			this.parts = value.split("\\.");
			checkParts(parts);
		}
	}

	public static TreePath create(String value) {
		if (value == null || value.length() == 0) {
			return EMPTY;
		}
		String[] parts = value.split("\\.");
		checkParts(parts);
		return new TreePath(value, parts);
	}

	private TreePath(String value, String[] parts) {
		this.value = value;
		this.parts = parts;
	}

	private static void checkParts(String[] parts) {
		for (String p : parts) {
			for (int j = 0; j < p.length(); j++) {
				if (!Character.isLetterOrDigit(p.charAt(j))) {
					throw new IllegalArgumentException("Invalid value for part: " + p + ". Only letters and digits allowed for labels");
				}
			}
		}
	}

	@Override
	public int hashCode() {
		return value.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		return obj instanceof TreePath && ((TreePath) obj).value.equals(this.value)
				|| obj instanceof String && obj.equals(this.value);
	}

	@Override
	public String toString() {
		return value;
	}

	public TreePath plus(TreePath other) {
		if (other == null) return this;
		if (this.value.length() == 0) {
			return other;
		} else if (other.value.length() == 0) {
			return this;
		}
		String[] values = Arrays.copyOf(this.parts, this.parts.length + other.parts.length);
		for (int i = 0; i < other.parts.length; i++) {
			values[this.parts.length + i] = other.parts[i];
		}
		return new TreePath(this.value + "." + other.value, values);
	}

	private static boolean compare(String[] left, String[] right, int count) {
		for (int i = 0; i < count; i++) {
			if (!left[i].equals(right[i])) return false;
		}
		return true;
	}

	public boolean isAncestor(TreePath other) {
		if (other == null) return false;
		if (this.value.equals(other.value)) return true;
		return this.parts.length < other.parts.length && compare(this.parts, other.parts, this.parts.length);
	}

	public boolean isDescendant(TreePath other) {
		if (other == null) return false;
		if (this.value.equals(other.value)) return true;
		return this.parts.length > other.parts.length && compare(this.parts, other.parts, other.parts.length);
	}
}
